internal/validator/common_validation.go

```diff
@@ -47,7 +47,7 @@ func CalculateMiningFeesRequired(size *bt.TxSize, fees *bt.FeeQuote) (uint64, er
 	return feesRequiredRounded, nil
 }
 
-func CommonValidateTransaction(policy *bitcoin.Settings, tx *bt.Tx) error {
+func CommonValidateTransaction(policy *bitcoin.Settings, tx *bt.Tx) *Error {
```

Comment:
I really really don't like that this returns `*Error`. I know this is defined in the same package, so it's internal in theory, but it's exported and looks like it can be used in other packages and it shouldn't IMO, because it's an error made just for validator.

Can we revert the name back to `ValidationError` / `ValidatorError`? 🙏

Reply:
@arkadiuszos4chain What is the difference between returning `error` or `*Error` here? 🤔 
I think the logic would be the same.

Reply:
@kuba-4chain I didn't change the name of the type. I think using a name like `ValidationError` or `ValidatorError` may not be idiomatic, as in client code it would look like `validator.ValidationError`. That said, it's not crucial for me, and if you think (and no one is against it), we could change the names.

@boecklim what I wanted to achieve with this change is to provide a clear understanding of when a client should wrap the error from these functions with `validator.Error`. Please take a look at the previous implementation:
``` go
func (v *DefaultValidator) ValidateTransaction(ctx context.Context, tx *bt.Tx, feeValidation validator.FeeValidation, scriptValidation validator.ScriptValidation) error { 
	if !v.IsExtended(tx) {
	       // Arek: note we wrap error here - why?
		return validator.NewError(fmt.Errorf("transaction is not in extended format"), api.ErrStatusTxFormat)
	}

	err := validator.CommonValidateTransaction(v.policy, tx)
	if err != nil {
	        // Arek: we do not wrap error - why?
		return err
	}

	switch feeValidation {
	case validator.StandardFeeValidation:
		if err := standardCheckFees(tx, api.FeesToBtFeeQuote(v.policy.MinMiningTxFee)); err != nil {
		          // Arek: no wrapping here - why?
			return err
		}
	case validator.NoneFeeValidation:
		// Do not handle the default case on purpose; we shouldn't assume that other types of validation should be omitted
	}

	// 12) The unlocking scripts for each input must validate against the corresponding output locking scripts
	if scriptValidation == validator.StandardScriptValidation {
		if err := checkScripts(tx); err != nil {
		        // Arek: here we are wrapping - why?
			return validator.NewError(err, api.ErrStatusUnlockingScripts)
		}
	}
	(...)
}

func standardCheckFees(tx *bt.Tx, feeQuote *bt.FeeQuote) error {
	feesOK, expFeesPaid, actualFeePaid, err := isFeePaidEnough(feeQuote, tx)
	if err != nil {
		return validator.NewError(err, api.ErrStatusFees)
	}

	if !feesOK {
		err = fmt.Errorf("transaction fee of %d sat is too low - minimum expected fee is %d sat", actualFeePaid, expFeesPaid)
		return validator.NewError(err, api.ErrStatusFees)
	}

	return nil
}

func checkScripts(tx *bt.Tx) error {
	for i, in := range tx.Inputs {
		prevOutput := &bt.Output{
			Satoshis:      in.PreviousTxSatoshis,
			LockingScript: in.PreviousTxScript,
		}

		if err := interpreter.NewEngine().Execute(
			interpreter.WithTx(tx, i, prevOutput),
			interpreter.WithForkID(),
			interpreter.WithAfterGenesis(),
		); err != nil {
			return fmt.Errorf("script execution failed: %w", err)
		}
	}

	return nil
}
```
Note that `standardCheckFees` and `checkScripts` both return error, but one of them needs wrapping (in the `ValidateTransaction` function) while the other does not. In my opinion, this leaks abstraction, forcing the developer to read and understand the body of the function to decide if he should wrap the returned error. With my change, the developer will know from the function definition alone: if it returns error, wrap it! If it returns validator.Error, return it as it is.

Another side effect is that in complex functions like `CommonValidation`, where we return several different errors from various branches, changing the definition to return `*Error` ensures that we return this type every time. With the previous implementation, there could be a situation where new code returns a standard error, which would be inconsistent and could lead to incorrect system behaviors.